programs/stake/src/stake_state.rs

```diff
@@ -705,6 +784,190 @@ pub fn redelegate(
+
+            let source_effective_stake = source_stake.delegation.stake;
+            let source_final_stake = source_effective_stake
+                .checked_sub(lamports)
+                .ok_or(InstructionError::InvalidArgument)?;
+
+            if source_final_stake != 0 && source_final_stake < minimum_delegation {
+                return Err(InstructionError::InvalidArgument);
+            }
+
+            let destination_meta = match destination_merge_kind {
+                MergeKind::FullyActive(destination_meta, mut destination_stake) => {
+                    if source_stake.delegation.voter_pubkey
+                        != destination_stake.delegation.voter_pubkey
+                    {
+                        return Err(StakeError::VoteAddressMismatch.into());
+                    }
+
+                    let destination_effective_stake = destination_stake.delegation.stake;
+                    let destination_final_stake = destination_effective_stake
+                        .checked_add(lamports)
+                        .ok_or(InstructionError::ArithmeticOverflow)?;
+
+                    if destination_final_stake < minimum_delegation {
+                        return Err(InstructionError::InvalidArgument);
+                    }
+
+                    merge_delegation_stake_and_credits_observed(
+                        &mut destination_stake,
+                        lamports,
+                        source_stake.credits_observed,
+                    )?;
+
+                    destination_account.set_state(&StakeStateV2::Stake(
+                        destination_meta,
+                        destination_stake,
+                        StakeFlags::empty(),
+                    ))?;
+
+                    destination_meta
+                }
+                MergeKind::Inactive(destination_meta, _, _) => {
+                    if lamports < minimum_delegation {
+                        return Err(InstructionError::InvalidArgument);
+                    }
+
+                    let mut destination_stake = source_stake;
+                    destination_stake.delegation.stake = lamports;
+                    destination_account.set_state(&StakeStateV2::Stake(
+                        destination_meta,
+                        destination_stake,
+                        StakeFlags::empty(),
+                    ))?;
+
+                    destination_meta
+                }
+                _ => return Err(InstructionError::InvalidAccountData),
+            };
+
+            if source_final_stake == 0 {
+                source_account.set_state(&StakeStateV2::Initialized(source_meta))?;
+            } else {
+                source_stake.delegation.stake = source_final_stake;
+                source_account.set_state(&StakeStateV2::Stake(
+                    source_meta,
+                    source_stake,
+                    StakeFlags::empty(),
```

Comment:
to be precise the idea is slightly trickier (although the behavior in code is still correct). the second case is an inactive stake becoming active (clearing any flags) but the first and third (this being the third) cases are an active stake that remains active, but we dont have the current stake flags from the mergekind because an active stake presently cannot have one

this will be one less thing to worry about with neostake. because the existing flag is part of the redelegate code, which is not being ported over, we will go back to a world with no flags